Change default log filename template to include map_index

ashb opened the pull request:
With the recently added LogTemplate mechanism old TIs will still use the format they had at creation time (with the change here to ensure that we create a LogTemplate row for the just-upgraded-in-place) so the logs can still be viewed in the UI.

Unfortuantely the only way I can find to have map_index only appear for mapped tasks, but also to retain full control of the filename to users was to put a conditional in the template, which I am not happy about.

And since it was now getting quite "deep" I have chosen to "label" the components in the "hive partition style".

~I'll need to test this change with Elasticsearch too~ ES working now.
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb committed:
Change default log filename template to include map_index

With the recently added LogTemplate mechanism old TIs will still use the
format they had at creation time (with the change here to ensure that we
create a LogTemplate row for the just-upgraded-in-place) so the logs can
still be viewed in the UI

And since it was now getting quite "deep" I have chosen to "label" the
components in the "hive partition style"

ashb commented on airflow/config_templates/config.yml:
```diff
@@ -598,7 +598,9 @@
       type: string
       example: ~
-      default: "{{{{ ti.dag_id }}}}/{{{{ ti.task_id }}}}/{{{{ ts }}}}/{{{{ try_number }}}}.log"
+      default: "dag={{{{ ti.dag_id }}}}/task_id={{{{ ti.task_id }}}}/run_id={{{{ ti.run_id }}}}/\
```
Thinking about this now we should probably change it to dag->run->task->map->try so that the order is least significant to most significant.

What do people think?

ashb replied:
Started discussion/lazy consensus on list https://lists.apache.org/thread/omt2zn02o3wcxtsg7zbr5yjxdo1offvd

ashb committed:
Make ti.set_context work for MappedOperators

The `except AttributeError` was _also_ catching more than just "this
handler doesn't have a set_context attribute", but also errors from
calling that function which lead to hard-to-track-down errors (missing
inlets/outlets)

I have also changed `get_template_context` to a side-effect-free
function, so it no longer mutates task.params!

ashb committed:
Change default template to put run_id before task_id

And instead of having to duplicate the default config value in to
configuration.py for the update process, change it to get the new
default value out of the loaded default_airflow.cfg

ashb committed:
Use `dag_id` not `dag` for consistency with other parameters

ashb committed:
Fix smart sensor logging

It has been broken in main for a while (it works fine in 2.2.x series),
but because we were catching _all_ AttributeErrors we never noticed.

ashb committed:
Fix tests

uranusjr commented on airflow/config_templates/config.yml:
```diff
@@ -598,7 +598,9 @@
       example: ~
-      default: "{{{{ ti.dag_id }}}}/{{{{ ti.task_id }}}}/{{{{ ts }}}}/{{{{ try_number }}}}.log"
+      default: "dag_id={{{{ ti.dag_id }}}}/run_id={{{{ ti.run_id }}}}/task_id={{{{ ti.task_id }}}}/\
+               {{%% if ti.map_index != -1 %%}}map_index={{{{ ti.map_index }}}}/{{%% endif %%}}\
```
It may be better to standardise the check to `map_index < 0` instead; IMO that’s more “logical” than specifically check for `-1`.

Of course, we shouldn’t have anything lower than -1, but in the very weird chance we hit one rouge value, it’s likely better to treat it as unmapped, since treating it as mapped would cause a crash later that’s difficult to debug.

kaxil replied:
worth updating description key too so users are aware about map_index and https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/logging-tasks.html too

ashb replied:
`map_index >= 0` in this case, but yeah good idea.

uranusjr commented on airflow/models/mappedoperator.py:
```diff
@@ -196,6 +196,9 @@ class MappedOperator(AbstractOperator):
     downstream_task_ids: Set[str] = attr.ib(factory=set, init=False)
 
+    inlets: list = attr.ib(factory=list)
+    outlets: list = attr.ib(factory=list)
```
I think these should be properties. Also probably don’t need to be declared in AbstractOperator?

ashb replied:
Properties looking at where? `parital_kwargs`?

uranusjr replied:
Yeah and populate the partial_kwargs key in `BaseOperator.partial` and `_TaskDecorator.map`.

uranusjr reviewed (commented):
Generally lgtm except two minor issues

ashb committed:
Code review changes

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -309,6 +309,8 @@ def map(self, **kwargs) -> XComArg:
         )
         partial_kwargs.setdefault("executor_config", {})
+        partial_kwargs.setdefault("inlets", [])
+        partial_kwargs.setdefault("outlets", [])
```
Should these be None?

ashb replied:
No, cos of this in BaseOperator init:

```        # Lineage
        self.inlets: List = []
        self.outlets: List = []
```

ashb replied:
Or not. I was wrong and these aren't needed at all actually.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -1836,7 +1838,7 @@ def get_template_context(
             params.update(task.params)
         if conf.getboolean('core', 'dag_run_conf_overrides_params'):
             self.overwrite_params_with_dag_run_conf(params=params, dag_run=dag_run)
-        validated_params = task.params = params.validate()
+        validated_params = params.validate()
```
Was setting `task.params` here unintended or simply unnecessary?

ashb replied:
I forget the error, but setting it here was giving some kind of error (I think from it being set _twice_ on the same in-memory object, once in the supervisor, and once in the actual runner.)

So I took the approach that `get_*` should never have any sideffects!

ashb committed:
Update Elasticsearch logging config to include map_index

Since the log_id is never really visible to users I have taken the
easier approach of just always including the map_index, even for
unmapped tasks.

kaxil commented on airflow/sensors/smart_sensor.py:
```diff
@@ -341,21 +340,21 @@ def _load_sensor_works(self, session=None):
+                sensor_works.append(SensorWork(si))
             except Exception:
-                self.log.exception("Exception at creating sensor work for ti %s", ti.key)
+                self.log.exception("Exception at creating sensor work for ti %s", si.ti_key)
```
Does `SensorInstance` have a `ti_key` attribute?

kaxil replied:
(I couldn't find `si.key` too)

ashb replied:
It does, yes.

ashb replied:
Er, no. Reverted this change. (I was thinking wrong class)

kaxil commented on docs/apache-airflow/logging-monitoring/logging-tasks.rst:
```diff
@@ -38,7 +38,12 @@ directory.
+The default pattern is followed while naming log files form tasks:
+
+- For normal tasks: ``dag_id{dag_id}/run_id={run_id}/task_id={task_id}/attempt={try_number}.log``.
+- For dynamically mapped tasks: ``dag_id{dag_id}/run_id={run_id}/task_id={task_id}/map_index={map_index}/attempt={try_number}.log``.
```
```suggestion
- For normal tasks: ``dag_id={dag_id}/run_id={run_id}/task_id={task_id}/attempt={try_number}.log``.
- For dynamically mapped tasks: ``dag_id={dag_id}/run_id={run_id}/task_id={task_id}/map_index={map_index}/attempt={try_number}.log``.
```

ashb replied:
🤦🏻

kaxil commented on docs/apache-airflow/logging-monitoring/logging-tasks.rst:
```diff
@@ -38,7 +38,12 @@ directory.
     For more information on setting the configuration, see :doc:`/howto/set-config`
 
-The following convention is followed while naming logs: ``{dag_id}/{task_id}/{logical_date}/{try_number}.log``
+The default pattern is followed while naming log files form tasks:
```
```suggestion
The default pattern is followed while naming log files for tasks:
```

ashb committed:
Apply suggestions from code review

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

ashb committed:
fixup! Code review changes

ashb committed:
fixup! Fix tests

ashb committed:
Revert unnecessary changes to smart_sensor
